internal/operator-controller/rukpak/bundle/config.go

```diff
@@ -0,0 +1,241 @@
+			return nil
+		},
+	}
+)
+
+// Config is a registry+v1 bundle configuration surface
+type Config struct {
+	// WatchNamespace is supported for certain bundles to allow the user to configure installation in Single- or OwnNamespace modes
+	// The validation behavior of this field is determined by the install modes supported by the bundle, e.g.:
+	// - If a bundle only supports AllNamespaces mode (or only OwnNamespace mode): this field will be unknown
+	// - If a bundle supports AllNamespaces and SingleNamespace install modes: this field is optional
+	// - If a bundle supports AllNamespaces and OwnNamespace: this field is optional, but if set must be equal to the install namespace
+	WatchNamespace string `json:"watchNamespace,omitempty"`
+}
+
+// ValidatedBundleConfigFromRaw returns a validated Config struct from the values given in rawConfig.
+// The applied validation will be determined by the install modes supported by the bundle
+func ValidatedBundleConfigFromRaw(rv1 RegistryV1, installNamespace string, rawConfig map[string]interface{}) (*Config, error) {
+	if len(rawConfig) == 0 {
+		return nil, nil
+	}
+
+	rawSchema := bundleConfigSchema(rv1, installNamespace)
+	customFormats := []*schemavalidation.Format{
+		dnsFormat,
+		notOwnNamespaceFmt(installNamespace),
+	}
+
+	if err := validateBundleConfig(rawSchema, customFormats, rawConfig); err != nil {
+		return nil, fmt.Errorf("invalid configuration: %v", err)
+	}
+
+	return toConfig(rawConfig)
+}
+
+// bundleConfigSchema generates a jsonschema used to validate bundle configuration
+func bundleConfigSchema(rv1 RegistryV1, installNamespace string) []byte {
+	// configure reflector
+	r := new(jsonschema.Reflector)
+	r.ExpandedStruct = true
+	r.AllowAdditionalProperties = false
+
+	// generate base schema
+	schema := r.Reflect(&Config{})
+
+	// apply bundle rawConfig based mutations for watchNamespace
+	configureWatchNamespaceProperty(rv1, installNamespace, schema)
+
+	// return schema
+	out, err := schema.MarshalJSON()
+	if err != nil {
+		panic(err)
+	}
+	return out
+}
+
+// configureWatchNamespaceProperty modifies schema to configure the watchNamespace config property based on
+// the install modes supported by the bundle marking the field required or optional, or restricting the possible values
+// it can take
+func configureWatchNamespaceProperty(rv1 RegistryV1, installNamespace string, schema *jsonschema.Schema) {
+	supportedInstallModes := sets.New[v1alpha1.InstallModeType]()
+	for _, im := range rv1.CSV.Spec.InstallModes {
+		if im.Supported && !unsupportedInstallModes.Has(im.Type) {
+			supportedInstallModes.Insert(im.Type)
+		}
+	}
+
+	allSupported := supportedInstallModes.Has(v1alpha1.InstallModeTypeAllNamespaces)
+	singleSupported := supportedInstallModes.Has(v1alpha1.InstallModeTypeSingleNamespace)
+	ownSupported := supportedInstallModes.Has(v1alpha1.InstallModeTypeOwnNamespace)
+
+	if len(supportedInstallModes) == 0 {
+		panic("bundle does not support any supported install modes")
```

Comment:
Panic-ing here seems drastic, unless there's some other precondition check that would guarantee that the bundle has at least one supported install mode.

If we do have a precondition check, maybe beef up the panic message with something like:
```suggestion
		panic("programmer error: bundle validation should have guaranteed support for at least one install mode, but found no supported install modes")
```

Reply:
I've refactored to return an error